[MonorepoBuilder] Fix autoloading for overlapping namespaces

mantiz opened the pull request:
There is a problem when two packages are using namespaces like `App` and `App\Foo`.

Currently only the namespace itself is checked to be contained within a `composer.json`, so when you have a namespace `App` it will be found in each `composer.json` using `App\<whatever>` and the path of the last matching `composer.json` will be taken.

I tried to fix this by adding quotation marks for the lookup, so partial matches won't be successful anymore.

TomasVotruba commented:
Hey, thanks for nice and clear PR.

Please add new test method instead of overriding old test, e.g. `testOverlapingPaths` to make clear they both work.

mantiz commented:
@TomasVotruba Thanks for your quick response. I reverted the previous commit and made a new one.

I modified the `getRelativeSourcePath` helper method, so I don't have to duplicate this part for another source directory.

Should I squash the commits?

TomasVotruba commented on packages/MonorepoBuilder/tests/ComposerJsonDecorator/AutoloadRelativePathComposerJsonDecorator/AutoloadRelativePathComposerJsonDecoratorTest.php:
```diff
@@ -51,14 +81,37 @@ private function getExpectedComposerJson(): array
             'autoload' => [
                 'psr-4' => [
-                    'App\\' => $this->getRelativeSourcePath() . '/src',
+                    'App\\' => $this->getRelativeSourcePath() . '/Source/src',
```
Why such a change?

Here the `Source` directory should be in the `$this->getRelativeSourcePath()` method. It does something different then the name tells.

TomasVotruba commented:
I see your comment now.
I'm such a case a new method with overlapping source

TomasVotruba commented:
You can squash them in the end. Thanks for such a high quality approach :+1:

mantiz commented:
Now I am using a dedicated method for the overlapping source path.

I'm not sure if I like it since now there is duplicated logic but I can't remove this without touching the method. ;-)

TomasVotruba commented:
Thank you :+1: 

Well, not everything duplicated have to be in separated method. I prefer readability and code that is chilly to read over purism and complicated constructions.

TomasVotruba commented:
You can squash it and it's ready to merge

TomasVotruba commented:
Btw, have you tried that command on this monorepo package? Is the result the same?

mantiz commented:
Ok, I squashed the commits now.

To be honest, I haven't tried to run the merge command before, but now I have and there are no changes.

TomasVotruba commented:
Let's :ship:  it !

mantiz commented:
Thank you very much and many thanks for this great tool. :bouquet:

TomasVotruba commented:
Thanks for nice word. Glad you enjoy it :+1: 

Do you have it in some open-sourced packages? I'd like to see it in action if you do

mantiz commented:
Not yet, just some private experimental stuff where I used to have many single repositories and I got overwhelmed by the management overhead. I will let you know when or if it will ever go public. ;-)

TomasVotruba commented:
Awesome :+1:  Look forward to it
